daliuge-engine/dlg/manager/node_manager.py

```diff
@@ -111,6 +111,108 @@ def _load(obj, callable_attr):
+    def __init__(self, max_workers: int):
+        self._max_workers = max_workers
+        self._thread_pool: ThreadPoolExecutor | None = None
+
+    def start(self, _rpc_endpoint):
+        logger.info("Initializing thread pool with %d workers", self._max_workers)
+        self._thread_pool = ThreadPoolExecutor(max_workers=self._max_workers)
+
+    def run_drop(self, app_drop: AppDROP) -> Future:
+        return self._thread_pool.submit(app_drop.run)
+
+    def close(self):
+        self._thread_pool.shutdown(wait=True)
+
+
+class NodeManagerProcessDropRunner(NodeManagerDropRunner):
+    # Process isolated properties - should only be accessed in @classmethods
+    # to ensure that they are global to a single process only
+    _rpc_client: typing.Optional[rpc.RPCClient]
+    _rpc_endpoint: typing.Tuple[str, int]
+
+    def __init__(self, max_workers: int):
+        self._max_workers = max_workers
+        self._process_pool: typing.Optional[ProcessPoolExecutor] = None
+
+    def start(self, rpc_endpoint):
+        logger.info("Initializing process pool with %d workers", self._max_workers)
+
+        self._process_pool = ProcessPoolExecutor(
+            max_workers=self._max_workers,
+            initializer=NodeManagerProcessDropRunner._setup_process,
+            initargs=(rpc_endpoint,),
+        )
+
+    @classmethod
+    def _setup_process(cls, rpc_endpoint):
+        signal.signal(signal.SIGTERM, signal.SIG_IGN)
+        signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+        cls._rpc_endpoint = rpc_endpoint
+        cls._rpc_client = rpc.RPCClient()
+
+        # We'll just let this be cleaned up when the process terminates
+        # (instead of the normal `shutdown()`)
+        cls._rpc_client.start()
+
+    def run_drop(self, app_drop: AppDROP):
+        inputs_proxy_info, outputs_proxy_info = NodeManagerProcessDropRunner._get_proxy_infos(app_drop)
+
+        # MP Queues pickle on a background thread - we need to ensure that we don't
+        # modify the input app_drop reference outside of the scope of this method
+        # so we create a shallow copy that can be pickled at an indeterminate point in the future
+        # See https://github.com/python/cpython/blob/v3.10.13/Lib/multiprocessing/queues.py#L95
+        copied_drop = copy.copy(app_drop)
+        copied_drop._inputs = collections.OrderedDict()
+        copied_drop._outputs = collections.OrderedDict()
+
+        return self._process_pool.submit(
+            NodeManagerProcessDropRunner._run_app_drop,
+            copied_drop, inputs_proxy_info, outputs_proxy_info,
+        )
+
+    @classmethod
+    def _run_app_drop(cls, app_drop, inputs_proxy_info, outputs_proxy_info):
+        cls._setup_drop_proxies(app_drop, inputs_proxy_info, outputs_proxy_info)
+        return app_drop.run()
+
+    @classmethod
+    def _setup_drop_proxies(
+        cls, app_drop: AppDROP, inputs_proxy_info, outputs_proxy_info
+    ):
+        app_drop._rpc_endpoint = cls._rpc_endpoint
```

Comment:
It's been a while since I saw this, but right now I'm having trouble understanding why do we need to set the `app_drop._rpc_endpoint` at all. It makes sense that we require inputs and outputs to have the information (which they do) so that we can create the proxies to them, but it doesn't seem necessary for the app drop itself that is running in the sub-process. Removing this would simplify a few places where this `rpc_endpoint` is being carried over only to for the sake of setting it here.

Reply:
Yeah again would need to search your past brain :stuck_out_tongue: A lot of these changes are from your earlier commits and I didn't really dive deeply into and mainly focused on using them to get the multiprocessing to work.

It looks like the change as you suggest it wouldn't be entirely straight forward though. Here we could just pass it into `ProxyInfo.from_data_drop()` directly, but `DynlibProcApp` also uses it as a flag to detect an RPC environment and we'd still need to pass a reference from somewhere into the `ProxyInfo.from_data_drop()` over there. Might be too hard basket for this MR?

Reply:
Again old me... yeah, it seems like I retained this purely to not affect `DynlibProcApp`? It definitely doesn't seem to be used here, nor should it.

Note that the suggestion I was making was to remove all this `cls._rpc_endpoint` handling through the `NodeManagerProcessDropRunner`, since _all_ drops (apps and data) are injected a `_rcp_endpoint` when they are created as part of a session that runs in a Node Manager. In other words, I think there's no need to receive an `rpc_endpoint` in `NodeManagerProcessDropRunner.start()`, nor to keep it as a class member.

What I'd propose would be to remove all the `rpc_endpoint` business from the `NodeManagerProcessDropRunner`, declare a `_rcp_endpoint` _class_ attribute in the base `Drop` class as `_rpc_endpoint: typing.Optional[typing.Tuple[str, int]] = None`, continue overriding it as you are [here](https://github.com/ICRAR/daliuge/pull/239/files#diff-fc98634ef2e1f098351efb5f47a6905aac4268d02a8b04207cd5f64f923b6ec9R390-R395), and check for `None` in `DynlibProcApp`. Does all of that make sense?

Also, I'd understand if you'd rather just see this work merged, this could be investigated/tackled separately.

Reply:
Ah cool, yeah that change seems reasonable, I'll give it a go (and verify all the tests pass locally since master build is broken).

Reply:
Alright, this works :+1: And the tests pass (locally)